src/foam/nanos/approval/ApprovableAwareDAO.js

```diff
@@ -0,0 +1,383 @@
+        { name: 'userId', type: 'long' }
+      ],
+      javaCode: `
+        RoleApprovalRequest request = (RoleApprovalRequest) req.fclone();
+        request.clearId();
+        request.setApprover(userId);
+        ((DAO) x.get("approvalRequestDAO")).put_(x, request);
+      `
+    },
+    {
+      name: 'fullSend',
+      type: 'void',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'request', type: 'RoleApprovalRequest' },
+        { name: 'obj', type: 'FObject' }
+      ],
+      javaCode:`
+      DAO requestingDAO;
+      Logger logger = (Logger) x.get("logger");
+
+      if ( request.getDaoKey().equals("approvableDAO") ) {
+        DAO approvableDAO = (DAO) x.get("approvableDAO");
+
+        Approvable approvable = (Approvable) approvableDAO.find(request.getObjId());
+
+        requestingDAO = (DAO) x.get(approvable.getDaoKey());
+      } else {
+        requestingDAO = (DAO) x.get(request.getDaoKey());
+      }
+
+      String modelName = requestingDAO.getOf().getObjClass().getSimpleName();
+
+      UserQueryService userQueryService = (UserQueryService) x.get("userQueryService");
+
+      User currentUser = (User) x.get("user");
+
+      List<Long> approverIds = userQueryService.getAllApprovers(getX(), modelName, currentUser);
+
+      if ( approverIds.size() <= 0 ) {
+        logger.error("No Approvers exist for the model: " + modelName);
+        throw new RuntimeException("No Approvers exist for the model: " + modelName);
+      }
+
+      if ( approverIds.size() == 1 && approverIds.get(0) == request.getInitiatingUser() ) {
+        logger.log("The only approver of " + modelName + " is the maker of this request!");
+        throw new RuntimeException("The only approver of " + modelName + " is the maker of this request!");
+      }
+
+      // makers cannot approve their own requests even if they are an approver for the model
+      // however they will receive an approvalRequest which they can only view and not approve or reject
+      // so that they can keep track of the status of their requests
+      sendSingleRequest(x, request, request.getInitiatingUser());
+      approverIds.remove(request.getInitiatingUser());
+
+      for ( int i = 0; i < approverIds.size(); i++ ) {
+        sendSingleRequest(getX(), request, approverIds.get(i));
+      }
+      `
+    },
+    {
+      name: 'put_',
+      javaCode: `
+      User user = (User) x.get("user");
+      Logger logger = (Logger) x.get("logger");
+
+      LifecycleAware lifecycleObj = (LifecycleAware) obj;
+      
+      if ( ! getIsEnabled() ){
+        if ( lifecycleObj.getLifecycleState() == LifecycleState.PENDING ){
+          lifecycleObj.setLifecycleState(LifecycleState.ACTIVE);
+        }
+        return super.put_(x,obj);
+      }
+
+      // system and admins override the approval process
+      if ( user != null && ( user.getId() == User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) {
```

Comment:
can this be a permission?  such as:   approvable.lifecyclestate.update.*

Reply:
Yes we can include it after this PR